### ISSUE TITLE
Dynamically add a Layer object to the map

### ISSUE THREAD

**Issue by HunteRoi** [alex3165/react-mapbox-gl#752]
Good morning everyone!

I am currently working on an application that needs to dynamically add draggable markers on the map. I have found the closed issue #223 but it is from 2017. As the library keeps getting updates, I am pretty sure things become different (because of API breaking changes for example).

I am looking for an example code and some explanations would not be too much if possible😄 
Also, maybe it would be great for this repo to have some kind of guide to understand how to build an application using your library - something like <https://discordjs.guide/> (please see <https://github.com/discordjs/guide>, @alex3165). 

Any input is appreciated! Thank you very much for your help 😀

**Comment by mklopets** [comment]
As for dynamically adding markers, you just need to plop in the `Marker` as a child of the react-mapbox-gl map – you can conditionally render them using React component state, for example. This library will take care of adding them to the map, then.

For the draggable part, react-mapbox-gl doesn't support dragging HTML Markers right now, but draggable Features are supported (see [switch-style demo](http://alex3165.github.io/react-mapbox-gl/demos)).

mapbox-gl-js (which this library just wraps) seems to support draggable HTML markers, so you could probably get that to work too by looking at their [example](https://docs.mapbox.com/mapbox-gl-js/example/drag-a-marker/). However, I'd go for the draggable Features solution outlined above if possible (i.e if it's just an image, text or basic shape you want to display – those can be done easily via symbol layers.

---

As for the question about documentation, I personally feel like the library's APIs are fairly well documented in the docs + demos. A lot of the problems you face when using react-mapbox-gl are actually problems with mapbox-gl itself, which also has a load of documentation and examples. I don't think there's any point in keeping all that information duplicated within this project.

For specific questions, I suggest first searching the issues here and/or googling – that's what I did when I first started using react-mapbox-gl and mapbox, and it worked out great. Always feel free to open issues or ask on Stack Overflow, if necessary. PRs are especially welcome for things you feel lack documentation.

**Comment by HunteRoi** [comment]
Thank you very much for your answer. I've managed to build the feature with Layer and Feature components although I am facing an issue with it.

It seems like the map does not acknowledge last updates of the variable containing all the coordinates added on the fly at some point. It looks like it rolls back to the previous state when dragging one point different from the previous one dragged to another location.
Do you have any thought on this? Here's a little demo :
https://i.gyazo.com/bbe921b44c4d6f7c8cf76bde1f55453e.mp4

____

It's not about a documentation but more about creating a guide for new people coming across the library. It is more useful than wasting time looking at here and there, testing out stuff and opening thousands of issues for an already answer question. This could benefit the project. See the discordjs library : they have an API documentation on one side and a guide on the other.
To resume, shifting the current documentation to a better one and adding a user guide to demonstrate the best ways of building basic and/or more specific features could be a huge plus to react-mapbox-gl.

**Comment by mklopets** [comment]
hmm, could you add repro-able code for this bug?

**Comment by HunteRoi** [comment]
Sure. https//github.com/oSoc19/opt-horison/tree/develop

**Comment by HunteRoi** [comment]
Hey guys!

Some :up: on this. I've managed to track down the problem to a falsy value in Javascript being the culprit, breaking the condition.
Things have been fixed, nothing has to do with the react-mapbox-gl library at all!

For reference, I have made a repository for this specific problem : 
https://github.com/HunteRoi/draggable-marker

This commit just fixes it :
https://github.com/HunteRoi/draggable-marker/commit/66a6422900316909d79f90ab5661d59197f88759

Thank you for your time @mklopets. Appreciated it.
